Reset allocation failure counter on condition

mhl-b opened the pull request:
Occasionally shard allocation might fail. Consistent failures will increment `failedAllocations` counter.  The `MaxRetryAllocationDecider` together with `index.allocation.max_retry(default=5)` will decide when to stop trying to allocate. We also persist failures counter in cluster state, restarting individual nodes wont reset counter.

Reaching max_retry count can create a problem when underlying issue for allocation is resolved, re-allocation will not happen. Currently we don't reset failures count automatically, only through REST API. 

In this change I introduce a new service that listens to cluster updates when new node joins, and triggers reroute with retry's reset.

idegtiarenko commented on server/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java:
```diff
@@ -92,6 +92,13 @@ public boolean indexRoutingTableChanged(String index) {
+     */
+    public boolean oldestIndexVersionChanged() {
+        return state.getMetadata().oldestIndexVersion() != previousState.getMetadata().oldestIndexVersion();
+    }
```
ClusterChangeEvent has a lot of usages. I suggest to make it a `private static boolean oldestIndexVersionChanged(ClusterChangeEvent event)` next to its only usage to make the interface of the ClusterChangeEvent as small as possible. This could be revisited if we see this is needed in multiple places.

mhl-b replied:
Ok. I noticed ClusterChangeEvent has some of the utility methods, but code is years old.

idegtiarenko commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java:
```diff
@@ -1276,6 +1276,15 @@ private void ensureMutable() {
+            if (shardRouting.unassignedInfo() == null) {
+                return false;
+            }
+            return shardRouting.unassignedInfo().getNumFailedAllocations() > 0;
```
According to https://github.com/elastic/elasticsearch/blob/6b419c13d318a14d33fdfdc9f209654db4035bcc/server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java#L108-L135, the initializing shard might have `unassignedInfo` as well. 
I do not think we need to reset error if it is attempting to start, only if it is stuck in `UNASSIGNED` state.

mhl-b replied:
My understanding that once shard is in initializing state we remove it from `unassignedShards`, isn't it? 

If yes then `unassignedShards.stream()` should have only unassigned. But this only a decision to trigger reroute with reset. The reroute will call`resetFailedCounter` and iterate over` unassignedIterator` when resets counters.

idegtiarenko commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationFailuresResetService.java:
```diff
@@ -0,0 +1,56 @@
+    @SuppressForbidden(reason = "legacy usage of unbatched task")
+    private void submitUnbatchedStateUpdateTask(ClusterStateUpdateTask task) {
+        clusterService.submitUnbatchedStateUpdateTask("reset-allocation-failures", task);
+    }
```
We would like to remove unbatched state update tasks eventually, lets avoid new usages.

To make a task batch-able you will need:
* A task queue: https://github.com/elastic/elasticsearch/blob/0a410b52161d060d69444d7253026db3a391bea2/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L161-L165
* A class/record that captures all the data required to compute the next cluster state: https://github.com/elastic/elasticsearch/blob/0a410b52161d060d69444d7253026db3a391bea2/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L327-L343
* An executor class that can process multiple task classes/records at once and apply them on a cluster state (here it will be enough to process a single tasks as they all will be identical?): https://github.com/elastic/elasticsearch/blob/0a410b52161d060d69444d7253026db3a391bea2/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L345-L353
* And finally submit the task: https://github.com/elastic/elasticsearch/blob/0a410b52161d060d69444d7253026db3a391bea2/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L253-L255

mhl-b replied:
I talked with David, he believes using un-batched version is ok here.

DaveCTurner replied:
Hmm now that I look at the implementation, I think we should consider batching here. Otherwise we could end up adding a sequence of nodes and then fruitlessly rerouting N times when once would be enough.

mhl-b replied:
Does it make sense to use RerouteService? But it will require changes there to support counter reset flag

DaveCTurner replied:
I think calling `AllocationService#reroute` ourselves is fine here.

idegtiarenko commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationFailuresResetService.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public record AllocationFailuresResetService(ClusterService clusterService, AllocationService allocationService)
+    implements
+        ClusterStateListener {
```
I like the idea of a separate listener/service as it allows us to select changes (in case we have more scenarios like node join) that should trigger error count resets. 

The con of this approach is that it introduces one more cluster state update in a series of changes once node joins.
I wonder if we can do the same as part of the existing cluster state update? Possibly in `org.elasticsearch.cluster.routing.allocation.AllocationService#disassociateDeadNodes` as it is called when node joins or leave the cluster.

DaveCTurner replied:
I'm pretty sure this'll be fine, bearing in mind that shard allocation failures should be very rare these days. I do like the flexibility and clarity of having this logic in one place rather than sprinkling it across the node-join and node-left executors (and who knows where else in future).

idegtiarenko commented:
You might want to take a look on

https://github.com/elastic/elasticsearch/blob/5f2fe57d5ee9c76b5190201d1cd2dccba453b5dd/server/src/internalClusterTest/java/org/elasticsearch/indices/IndicesLifecycleListenerIT.java#L130-L143

Above integration test creates an index and exhaust its error count by making it impossible to move.
Similar approach could be used to test this change.

mhl-b committed:
add reset service and unit test

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

mhl-b commented:
@idegtiarenko, I updated code and added unit test. Still working on the integ test.

I tried to include change into NodeJoinExecutor rather than create new service as we talked offline. TLDR, I think it still requires another cluster state update, so it wont save an extra call. So I'm back to implementation with service.

idegtiarenko commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationFailuresResetServiceTests.java:
```diff
@@ -0,0 +1,165 @@
+        var changeEvent = new ClusterChangedEvent("", state, state);
+        var listener = new SubscribableListener<ClusterState>();
+        resetService.processEvent(changeEvent, listener);
+        var resultState = safeAwait(listener);
```
Consider using `DeterministicTaskQueue`:

```
var taskQueue = new DeterministicTaskQueue();
var threadPool = taskQueue.getThreadPool();
...
clusterChanged(event);
taskQueue.runAllRunnableTasks();
...
```
This way you can guarantee all tasks are executed without exposing a listener for test only purposes.

mhl-b replied:
how do I assert result? through cluster service?

idegtiarenko replied:
It should be accessible via `ClusterService#state`

mhl-b replied:
Couldnt make cluster service work together with deterministic queue. The reroute code path always runs after unit test code is done. After few hours of browsing other tests and try/err, I decided to use `TestThreadPool` and `assertBusy` block that will eventually get correct cluster state.

mhl-b committed:
update unit test

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationFailuresResetService.java:
```diff
@@ -0,0 +1,62 @@
+ * new node join covers a variety of use cases, such as rolling update, version
+ * (and other settings), node restarts.
+ */
+public class AllocationFailuresResetService implements ClusterStateListener {
```
Do we subscribe this listener to the cluster service somewhere? Sorry if I'm missing the line that does that. It'd be good to have an integ test to demonstrate that it is all wired in correctly as well as the unit tests that show that the service works in isolation.

mhl-b replied:
Will add integ test, it will catch missing subscription

DaveCTurner reviewed (commented):
The logic looks good, I left a couple of comments about the surrounding code tho.

mhl-b committed:
address feedback, remove service

mhl-b committed:
Merge remote-tracking branch 'upstream/main' into reset-failure-counter

mhl-b commented:
@DaveCTurner @idegtiarenko 
I've updated PR with batched executor and added integration test. I also decided to remove AllocationFailuresResetService and make it a method in AllocationService, which creates and adds listener to cluster service. I found it's cleaner and less verbose.

idegtiarenko commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -559,6 +562,25 @@ private void allocateExistingUnassignedShards(RoutingAllocation allocation) {
+    public void addAllocFailuresResetListenerTo(ClusterService clusterService) {
+        // batched cluster update executor, runs reroute once per batch
+        // set retryFailed=true to trigger failures reset during reroute
+        var taskQueue = clusterService.createTaskQueue("reset-allocation-failures", Priority.URGENT, (batchCtx) -> {
```
Lest use `Priority.NORMAL` here.

mhl-b committed:
change task priority to normal

mhl-b committed:
Merge remote-tracking branch 'upstream/main' into reset-failure-counter
